products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -0,0 +1,172 @@
+import WinSDK
+#else
+import Glibc
+#endif
+
+import Foundation
+
+// We don't need this import if we're building
+// this file as part of the llbuild framework.
+#if !LLBUILD_FRAMEWORK
+import llbuild
+#endif
+
+public typealias KeyID = UInt64
+public typealias KeyType = String
+
+/// Defines the result of a call to fetch all keys from the database
+/// Wraps calls to the llbuild database, but all results are fetched and available with this result
+public class BuildDBKeysResult {
+    /// Opaque pointer to the actual result object
+    private let result: OpaquePointer?
+    
+    fileprivate init() {
+        self.result = nil
+    }
+    
+    fileprivate init(result: OpaquePointer) {
+        self.result = result
+    }
+    
+    private lazy var _count: Int = result.map { Int(llb_database_result_keys_get_count($0)) } ?? 0
+    
+    /// Get the key for a given key identifier (might be unique)
+    public func getKey(at index: KeyID) -> KeyType {
+        assert(result != nil, "Can't get key without fetching the data first.")
+        
+        var data = llb_data_t()
+        withUnsafeMutablePointer(to: &data) { ptr in
+            llb_database_result_keys_get_key_at_index(self.result, index, ptr)
+        }
+        defer { data.data?.deallocate() }
+        return stringFromData(data)
+    }
+    
+    deinit {
+        llb_database_destroy_result_keys(result)
+    }
+}
+
+extension BuildDBKeysResult: Collection {
+    public typealias Index = KeyID
+    public typealias Element = KeyType
+    
+    public var startIndex: Index {
+        return 0 as UInt64
+    }
+    
+    public var endIndex: Index {
+        return UInt64(self._count) + startIndex
+    }
+    
+    public subscript(index: Index) -> Iterator.Element {
+        guard (startIndex..<endIndex).contains(index) else {
+            fatalError("Index \(index) is out of bounds (\(startIndex)..<\(endIndex))")
+        }
+        return getKey(at: index)
+    }
+    
+    public func index(after i: KeyID) -> KeyID {
+        return i + 1
+    }
+}
+
+extension BuildDBKeysResult: CustomReflectable {
+    public var customMirror: Mirror {
+        let keys = (startIndex..<endIndex).map { getKey(at: $0) }
+        return Mirror(BuildDBKeysResult.self, unlabeledChildren: keys, displayStyle: .collection)
+    }
+}
+
+/// Private class for easier handling of out-parameters
+private class MutableStringPointer {
+    var ptr = llb_data_t()
+    init() { }
+    
+    deinit {
+        ptr.data?.deallocate()
+    }
+    
+    var msg: String? {
+        guard ptr.data != nil else { return nil }
+        return stringFromData(ptr)
+    }
+}
+
+/// Database object that defines a connection to a llbuild database
+public final class BuildDB {
+    
+    /// Errors that can happen when opening the database or performing operations on it
+    public enum Error: Swift.Error {
+        /// If the system can't open the database, this error is thrown at init
+        case couldNotOpenDB(String)
```

Comment:
Can you please add labels to the parameters here to make it clear what the purpose of the argument is?